Enable WP_Filesystem_VIP

sboisvert opened the pull request:
This commit enables the WP_Filesystem_VIP class by adding the filters that force all wp_filesystem() calls to return an instance of WP_Filesystem_VIP which then transparently handles local file modifications as well as direct use of $wp_filesystem

### Test plan

1) Does the standard media upload still work after update?
2) Does running:
```
$creds = request_filesystem_credentials( site_url()
wp_filesystem($creds);
```
Give you back an instance of the `WP_Filesystem_VIP` class
3) Can you reach and write files using the `wp_filesystem($creds);` to `/tmp/` and to `wp_get_upload_dir()['basedir']`

sboisvert commented on files/init-filesystem.php:
```diff
@@ -0,0 +1,58 @@
+ *         new Automattic\VIP\Files\WP_Filesystem_VIP_Uploads( $api_client ),
+ *         new WP_Filesystem_Direct( null ),
+ *     ] );
+ *     $filesystem->get_contents( '...' );
```
Should we replace them with an example like this:

global $wp_filesystem;
if ( ! is_a( $wp_filesystem, 'WP_Filesystem_Base') ){
$creds = request_filesystem_credentials( site_url() );
wp_filesystem($creds);
}
$wp_filesystem->put_contents( wp_get_upload_dir()['basedir'] . '/test.txt', 'this is a test file');

to give folks a sense of how to use it?

mjangda replied:
Yep, that's a good idea.

Aside, is calling `request_filesystem_credentials` necessary? Would be nice to do that automatically instead, maybe something our "stub" class can do?

```
class WP_Filesystem_VIP extends Automattic\VIP\Files\WP_Filesystem_VIP {
    public function __construct( $args ) {
        // Make it easier to instantiate our VIP class.
        // Handle the "credential request" so clients can just call `WP_Filesystem()` and be done.
        if ( null === $args ) {
            $args = request_filesystem_credentials( site_url() );
        }
        parent::__construct( $args );
    }
}
```

sboisvert commented on files/init-filesystem.php:
```diff
@@ -0,0 +1,58 @@
+class WP_Filesystem_VIP extends Automattic\VIP\Files\WP_Filesystem_VIP {}
+
+add_filter( 'filesystem_method', function( $method, $args, $context, $allow_relaxed_file_ownership ) {
+	return 'VIP'; // All VIP, all the time
```
We could, what would be the benefit?

mjangda replied:
Repeating hard-coded strings is generally a non-ideal practice? Constants provide some safety. More here https://stackoverflow.com/a/2953613

sboisvert replied:
Right, I guess in this context I don't see it's usage expanding and creating a constant for it seems a bit like overkill and I'm not sure I see the full benefits. That being said, I'm not against it, it just feels unnecessary in this use case :)

mjangda commented:
@sboisvert are there are any specific core flows / code paths that we should test to verify that this change doesn't break anything critical?

sboisvert commented:
I've added the three things I tested for when double checking this PR in the top of the PR.
The first one checks to ensure our current filters for uploading media still work. The other two look more into the direct functionality of the new class.

(I accidentally edited your comment for some reason while posting this and reverted it back)

mjangda commented:
One blocker here is that we need to appropriately handle local environments which will not be able to make requests to the Files Service.

For now, the best option would be to fallback to `WP_Filesystem_Direct`.

sboisvert commented:
@mjangda Good call, I've just updated it to do that.

mjangda commented:
A couple of possible improvements in future iterations:

- Remove the need to call is calling `request_filesystem_credentials`. Would be nice to do that automatically instead, maybe something our "stub" class can do:

```
class WP_Filesystem_VIP extends Automattic\VIP\Files\WP_Filesystem_VIP {
    public function __construct( $args ) {
        // Make it easier to instantiate our VIP class.
        // Handle the "credential request" so clients can just call `WP_Filesystem()` and be done.
        if ( null === $args ) {
            $args = request_filesystem_credentials( site_url() );
        }
        parent::__construct( $args );
    }
}
```

- For local development, have a custom Filesystem class that fails all the write actions.

sboisvert commented:
> Remove the need to call is calling request_filesystem_credentials. Would be nice to do that automatically instead, maybe something our "stub" class can do:

I thought about doing a wrapper that did this, but I decided against it because I feel like many of the times this will be used will be in community plugins. If we suggest and offer a wrapper it'll means that people will do stuff like:

```
if ( function_exists( wpcom_vip_get_filesystem() ){
$wp_filesystem = $wpcom_vip_get_filesystem()
}else{
[coreway]
}
```

Which leads to custom code for the platform vs doing it the same as core everywhere.


> For local development, have a custom Filesystem class that fails all the write actions (unless it's the tmp path).

https://github.com/Automattic/vip-go-mu-plugins/blob/master/files/class-wp-filesystem-vip.php#L44

Should be taking care of this.

mjangda commented:
> I thought about doing a wrapper that did this, but I decided against it because I feel like many of the times this will be used will be in community plugins.

I wasn't suggesting a wrapper; more along the lines of this: than https://github.com/Automattic/vip-go-mu-plugins/pull/956#discussion_r209022343

```
class WP_Filesystem_VIP extends Automattic\VIP\Files\WP_Filesystem_VIP {
    public function __construct( $args ) {
        // Make it easier to instantiate our VIP class.
        // Handle the "credential request" so clients can just call `WP_Filesystem()` and be done.
        if ( null === $args ) {
            $args = request_filesystem_credentials( site_url() );
        }
        parent::__construct( $args );
    }
}
```

sboisvert commented:
I think I understand now. Is the idea that this is currently how plugins are using it? From reading the documentation, while it would work, would it make it less portable? Is it a best practice to do the request_filesystem_credentials() call before initializing it? It was my understanding that it was so here we'd be encouraging folks to use `WP_Filesystem()` over doing the credentials request first.

All that being said, I don't have a strong preference either way so if you feel it's useful I'm happy to do a PR for it.
